### PR TITLE
<feat>(pbft): add rotate proposal check, add working sealer manager ut.

### PR DESCRIPTION
**Pull request by kyonRay**

### CODE REVIEW

**Review by morebtcg — APPROVED** [review]

File: bcos-executor/src/precompiled/common/WorkingSealerManagerImpl.cpp [hunk 1]
```diff
@@ -46,13 +45,13 @@ void WorkingSealerManagerImpl::rotateWorkingSealer(
...
+        getConsensusNodeListFromStorage(_executive, _callParameters->m_origin);
         checkVRFInfos(parentHash, _callParameters->m_origin);
     }
-    catch (std::exception const& e)
+    catch (protocol::PrecompiledError const& e)
     {
-        PRECOMPILED_LOG(WARNING) << LOG_DESC("rotateWorkingSealer failed for checkVRFInfos failed")
-                                 << LOG_DESC("notifyNextLeaderRotate now");
+        PRECOMPILED_LOG(WARNING) << LOG_DESC(
+            "rotateWorkingSealer failed for checkVRFInfos failed, notifyNextLeaderRotate now");
```
Comment (bxq2011hust) [inline bcos-executor/src/precompiled/common/WorkingSealerManagerImpl.cpp:54]:
```suggestion
            "rotateWorkingSealer checkVRFInfos failed, notifyNextLeaderRotate now");
```

File: bcos-executor/src/precompiled/common/WorkingSealerManagerImpl.cpp [hunk 2]
```diff
@@ -215,6 +211,31 @@ void WorkingSealerManagerImpl::getConsensusNodeListFromStorage(
         }
     }
     m_consensusNodes.swap(consensusNodeList);
+    // check origin: the origin must be among the workingSealerList
+    if (!m_workingSealer.empty()) [[likely]]
```
Comment (bxq2011hust) [inline bcos-executor/src/precompiled/common/WorkingSealerManagerImpl.cpp:215]: getConsensusNodeListFromStorage should not check origin

File: bcos-executor/test/unittest/libprecompiled/PreCompiledFixture.h [hunk 3]
```diff
@@ -287,8 +287,9 @@ class PrecompiledFixture : public TestPromptFixture
             [m_blockHeader = bcostars::BlockHeader()]() mutable { return &m_blockHeader; });
         blockHeader->setNumber(blockNumber);
 
-        std::vector<bcos::protocol::ParentInfo> parentInfos{
-            {blockHeader->number() - 1, h256(blockHeader->number() - 1)}};
+        bcos::protocol::ParentInfo p{
+            .blockNumber = blockNumber - 1, .blockHash = h256(blockNumber - 1)};
+        std::vector<bcos::protocol::ParentInfo> parentInfos{p};
```
Comment (bxq2011hust) [inline bcos-executor/test/unittest/libprecompiled/PreCompiledFixture.h:292]: from my opinion left is better

File: bcos-framework/bcos-framework/ledger/LedgerTypeDef.h [hunk 4]
```diff
@@ -86,6 +84,24 @@ constexpr static std::string_view SMALLBANK_TRANSFER{"/tables/smallbank_transfer
 constexpr static std::string_view SYS_CODE_BINARY{"s_code_binary"};
 constexpr static std::string_view SYS_CONTRACT_ABI{"s_contract_abi"};
 
+enum ConsensusType : uint32_t
+{
+    PBFT_TYPE = 1,
+    RPBFT_TYPE = 2,
+};
+
+inline uint32_t consensusTypeFromString(std::string_view consensus)
```
Comment (bxq2011hust) [inline bcos-framework/bcos-framework/ledger/LedgerTypeDef.h:93]: this function seems not used?

File: bcos-pbft/bcos-pbft/pbft/engine/PBFTEngine.cpp [hunk 5]
```diff
@@ -775,6 +776,55 @@ bool PBFTEngine::checkProposalSignature(
         nodeInfo->nodeID(), _proposal->hash(), _proposal->signature());
 }
 
+bool PBFTEngine::checkRotateTransactionValid(
```
Comment (bxq2011hust) [inline bcos-pbft/bcos-pbft/pbft/engine/PBFTEngine.cpp:779]: should we do this in **RPBFTEngine**?

Comment (kyonRay) [reply bcos-pbft/bcos-pbft/pbft/engine/PBFTEngine.cpp:779]: for convenience of dynamic change consensus type, this function will return directly when consensus type is pbft.

**Review by morebtcg — APPROVED** [review]

**Review by bxq2011hust — APPROVED** [review]

**Pull request merged**